### DISCUSSION TITLE
进一步完善x-reactions的扩展能力

### DISCUSSION THREAD

**Discussion by janryWang** [alibaba/formily#1629] [Ideas]
最近在搞表单设计器，但是发现异步数据源这块怎么在schema描述却是个问题，总不能全部走纯函数表达式形式，那样太不方便存储，而且无法做到可视化回显，所以，想进一步增强一下x-reactions的能力，比如：

```
x-reactions={[
 {
   type:"dataSource", //同步数据源形态，支持依赖数据动态切换数据源
   dependencies:["aa"],
   conifgs:[
    {
       when:"{{$deps[0]}}",
       dataSource:[
         {label:"",value:""}
       ]
     }
   ]
 },
 {
   type:"asyncDataSource", //异步数据源形态，支持依赖数据动态请求传参
   dependencies:["xxx"],
   configs:{
     action:"//aa.bb.cc/{id}/xx",
     method:"POST",
     searchKey:"",
     labelKey:"",
     valueKey:"",
     childrenKey:"",
     params:{
       id:"{{$deps[0]}}"
     },
   }
 },
 {
   type:"likange", //标准联动形态，作为默认类型
   dependencies:[],
   when:"",
   fulfill:{
   
   },
   otherwise:{
   
   }
 },
 {
   type:"formula", //公式形态
   configs:{
     formula:"MIN(aa.bb.cc,cc.dd.ee)"
   }
 }
]}
```

然后Schema静态方法给一个专门的注册器，其实还是registerPatch的变体

```
Schema.registerReactionHandler('dataSource',(configs,field)=>{
  //...
})

```

**Comment by janryWang** [comment]
然后，要支持ReactionHandler的定制，其实就涉及需要维护局部状态的能力了，这就牵扯到 https://github.com/alibaba/formily/discussions/1590

**Comment by liuweiGL** [comment]
我觉得可以更开放一点，引入 api 管理系统，比如集成 pont 类库的能力；直接配置一个后台的 swagger 地址生成 api 模块，然后组件可以选择一个 api 作为数据源。

**Reply by janryWang** [reply]
这个偏上层应用了，咱们主要讨论api

**Reply by nekic** [reply]
如果从整个系统的构建角度来说，我是比较倾向引入一个 所谓的 api 管理模块或者数据源管理，就在 designable 上实现，然后 formily 需要使用动态数据的时候，就仅仅需要指明是那个 [数据源] , 由每个 Field 主动注册一个 dataSourceFetcher

**Reply by nekic** [reply]
如果就单纯从formily 的角度上讲，我也认为应该把 这种动态的东西隔离出去，因为 api 本身就太复杂了，查询参数、会话认证、结果集转换等等，所以，在每个 Field 上实现一个所谓的 dataSourceFetcher，由外部注入，不管其中实现，只要 dataSourceFetcher 能拿到真正的 处理好的 dataSource 就行了

**Comment by feynmx** [comment]
🤔,可不可以有个datasource的管理模块,负责注册数据源,formily 关联datasource,接下来就是传参取数.这样formily并不关心fetch的配置和逻辑,这种也利于系统集成

**Comment by yanchangyou** [comment]
还有一种方式，只配置一个函数名称，查询逻辑由使用者自己实现

**Reply by janryWang** [reply]
这种现在就支持x-reactions:"{{fn}}"

**Comment by janryWang** [comment]
算了， 不打算搞这么复杂扩展能力了，不管是联动还是异步数据源，直接上低代码更快

**Reply by wsxe9988** [reply]
嘿嘿，这样的扩展能力，在designeable 有这个 一些简单的会方便不少，复杂的还是写代码更快
